KAFKA-15417  flip joinSpuriousLookBackTimeMs and emit non-joined items

VictorvandenHoven opened the pull request:
Flipped joinSpuriousLookBackTimeMs, because they were not correct.
Moved the "emit non-joined items"-logic  after the "joined items"-logic instead of before, because only then you know whether to emit or not.
Introduced the windowsAfterIntervalMs-field that is used to find if emitting records can be skipped.


Now the  KStreamKStreamLeftJoinTest tests leftJoin with  .after(ZERO).
(Invalid) unitTests have been modified in KStreamKStreamLeftJoinTest and KStreamKStreamOuterJoinTest according to new logic.

VictorvandenHoven committed:
KAFKA-15417  flip joinSpuriousLookBackTimeMs and emit non-joined items
after the joined items in stead of before.

VictorvandenHoven committed:
KAFKA-15417  Checkstyle issues

VictorvandenHoven committed:
KAFKA-15417  layout

VictorvandenHoven committed:
KAFKA-15417  force build

VictorvandenHoven committed:
KAFKA-15417  force build

VictorvandenHoven committed:
Merge branch 'apache:trunk' into KAFKA-15417-flip-joinSpuriousLookBackTimeMs

mjsax commented:
Thanks for the PR. I did not forget about it (sorry for the wait; very busy times...).

> Moved the "emit non-joined items"-logic after the "joined items"-logic instead of before, because only then you know whether to emit or not.

Can you elaborate? Not sure if I can follow?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoinTest.java:
```diff
@@ -609,17 +842,18 @@ public void testOrdering() {
-                new KeyValueTimestamp<>(0, "A0+null", 0L),
-                new KeyValueTimestamp<>(1, "A1+a1", 110L)
+                new KeyValueTimestamp<>(1, "A1+a1", 110L),
+                new KeyValueTimestamp<>(0, "A0+null", 0L)
```
It seems your change to push out "left/outer" join result at the end of `process()` instead of at the beginning leads to this revers output, what implies we now emit data out-of-order. -- Hence, it seems better to keep the code as-is?

VictorvandenHoven replied:
Can you point me to the ordering-rules in case of a  left- outer- inner- join, then I can have a look.

lihaosky replied:
This is because previously we look at outer store and then join. But this change make it we join first and then look at outer store. The ts in outer store and other store is hard to reason. If we change the ts of 0 to be 100 and ts of 1 to be 50, the original test would still produce 0 first which has larger ts... So unless we compare the ts of join and outer at the same time when we output, we can't guarantee the order of ts when output.

VictorvandenHoven replied:
Agreed

mjsax replied:
Just cycling back to this PR.

> If we change the ts of 0 to be 100 and ts of 1 to be 50

For this case, we have out-of-order input data, so it's ok if we produce out-of-order results, right? But if input data is in-order, should we not try to produce ordered output, too?

In the end, my question is: why is it required to change the order of the code and now check the "outer" store last? From an order perspective, it seems better to check the "outer" store first, because it would emit record with lower timestamp first?

VictorvandenHoven replied:
What we have seen is that you could emit both a null-joined record and a joined-record if you first check the outerjoinstore  (via emitNonJoinedOuterRecords()) and only after that, a record from the otherWindowStore. 
This behaviour is not what you want.
Therefore you should first find out if there is going to be a joined record and if there is, you have to nullify this in the relevant outerjoinstore.

mjsax replied:
I agree that we should not produce a null-joined record and joined-record. Can you elaborate on the case when (and why) this could exactly happen? It sounds like a bug in `emitNonJoinedOuterRecords()` to me.

When we check the `outerStore` we should only emit null-joined-records that are "expired", ie, which belong to already closed windows, and thus, the order in which we check should not matter. It sounds like as if we might incorrectly emit null-joined records of window that are not closed yet (would be a bug) or for windows which are already closed but we emit an incorrect join-record.

Given that we identified that we indeed have a bug with regard to "late records" and that we don't respect the grace-period correctly, I would assume it's the same root cause.

Thus, I am now wondering if we (1) should merge this PR as-is (and fix the order back in a follow PR @florin-akermann is doing), or (2) revert this change in this PR right away, or (3) first try to merge Florin's PR and re-evaluate this change afterwards?

Option (1) seems to be the least desirable to me. I am happy with both (2) or (3).

Thoughts?

VictorvandenHoven replied:
When we change the order in which we check back to as it was, only the ordering tests will fail. So there seems to be no bug (anymore).
I think that in a previous version of this PR (before checking both the left and right in the outerjoin) we saw that it did matter and a left-join test failed. But this has been solved now with the getOuterJoinLookBackTimeMs().
I think option 2 and option 3 will do fine.

mjsax replied:
Seems this PR is basically ready for merging, so it might be faster to go with option (2), and revert changing the order in this PR and we can merge it. -- Of course, I would want to make a final pass after the change, to check the testing code again to verify correctness.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamOuterJoinTest.java:
```diff
@@ -790,9 +790,7 @@ public void shouldNotEmitLeftJoinResultForAsymmetricBeforeWindow() {
-            processor.checkAndClearProcessResult(
-                new KeyValueTimestamp<>(1, "null+a1", 102L)
-            );
+            processor.checkAndClearProcessResult();
```
Oh dear... the comment already says "should not produce any items"... Embarrassing that we still did not catch this bug originally...

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamOuterJoinTest.java:
```diff
@@ -878,7 +878,9 @@ public void shouldNotEmitLeftJoinResultForAsymmetricAfterWindow() {
 
-            // push one item to the first stream; this should produce one full-join item
+            // push one item to the first stream;
+            // this should produce one full-join item;
```
I just realize that it should be `inner-join item` (not `full-join item`) -- could we fix this across the broad in this PR as side cleanup?

VictorvandenHoven commented:
> Thanks for the PR. I did not forget about it (sorry for the wait; very busy times...).
> 
> > Moved the "emit non-joined items"-logic after the "joined items"-logic instead of before, because only then you know whether to emit or not.
> 
> Can you elaborate? Not sure if I can follow?

Sure.
If not-joined record is emitted at window-close, but there is also a joined-record available at that time, then both not-joined and joined would be emitted wouldn't they?
In the joined-item-logic the not-joined record in the outerJoinStore is nullified (correctly) but it has already been emitted before.


Or am I mistaken?

lihaosky commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoinTest.java:
```diff
@@ -436,6 +436,239 @@ public void testRightNonJoinedRecordsAreNeverEmittedByTheRightProcessor() {
+        joined = stream1.leftJoin(
+            stream2,
+            MockValueJoiner.TOSTRING_JOINER,
+            JoinWindows.ofTimeDifferenceWithNoGrace(ofMillis(100)).after(ofMillis(0)),
```
nit: `ofMillis(0)` -> `ZERO`?

lihaosky commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoinTest.java:
```diff
@@ -436,6 +436,239 @@ public void testRightNonJoinedRecordsAreNeverEmittedByTheRightProcessor() {
+            for (int i = 0; i < expectedKeys.length; i++) {
+                inputTopic1.pipeInput(expectedKeys[i], "B" + expectedKeys[i], time + i);
+            }
+            processor.checkAndClearProcessResult(
+                    new KeyValueTimestamp<>(0, "B0+null", 1000L),
+                    new KeyValueTimestamp<>(1, "B1+null", 1001L),
+                    new KeyValueTimestamp<>(2, "B2+null", 1002L)
+            );
+        }
```
Noob question: why do we have output here? The time difference is `100ms`, should we only output these three if we got an event with time `1103`? Maybe I'm missing something

VictorvandenHoven replied:
No,it is a leftjoin with joinwindows 100ms before and 0ms after.
So directly at timestamp of B0, B1 and B2 the not-joined record will be emitted.

lihaosky reviewed (commented):
Thanks for the fix!

VictorvandenHoven committed:
Merge branch 'apache:trunk' into KAFKA-15417-flip-joinSpuriousLookBackTimeMs

VictorvandenHoven committed:
KAFKA-15417  nit: ofMillis(0) -> ZERO?

lihaosky commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
```diff
@@ -228,7 +231,7 @@ private void emitNonJoinedOuterRecords(
 
                     // Skip next records if window has not closed
-                    if (timestamp + joinAfterMs + joinGraceMs >= sharedTimeTracker.streamTime) {
+                    if (sharedTimeTracker.minTime + windowsAfterIntervalMs  + joinGraceMs >= sharedTimeTracker.streamTime) {
```
I'm not sure about this part. If the `value` has right value and this is on right side, why don't we check with `windows.beforeMs` to see if it expires? Not sure why `windowsAfterIntervalMs` is always set to `windows.afterMs`.

VictorvandenHoven replied:
I think you are right. 

It depends on the side of the outerJoin-value whether we should check with _windowsAfterMs_ or _windowsBeforeMs_.

Since the outerJoin-store can contain both left-sided and right-sided outerJoin-values, we should check with _windowsAfterMs_ when outerJoin-value is left-sided and we should check _windowsBeforeMs_ when outerJoin-value is right-sided.

Also, we can not break the emitNonJoinedOuterRecords-while-loop until we are completely sure that there are no more left-sided and right-sided outerJoin-values available to emit. 

For example, if we find out that we can skip a left-sided outerJoin-value, since the window for this value has not yet been closed, there can still be a right-sided outerJoin-value that must be emitted.

In the future maybe some leftSide/rightSide-flags could be introduced that indicate whether we have put left-sided or right-sided outerJoin-values in the outerJoin-store.  So that with a leftJoin() we could break the emitNonJoinedOuterRecords-while-loop earlier in order to gain time.

VictorvandenHoven committed:
KAFKA-15417  Added the outerJoinSide dependent logic to determine
emitting non-joined records and removed the obsolete
joinSpuriousLookBackTimeMs (= joinAfterMs)

VictorvandenHoven committed:
KAFKA-15417  checkStyle violations

lihaosky commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
```diff
@@ -60,24 +61,21 @@ class KStreamKStreamJoin<K, V1, V2, VOut> implements ProcessorSupplier<K, V1, K,
-                       final TimeTrackerSupplier sharedTimeTrackerSupplier) {
+    KStreamKStreamJoin(final boolean isLeftSide, final String otherWindowName, final JoinWindowsInternal windows,
+            final ValueJoinerWithKey<? super K, ? super V1, ? super V2, ? extends VOut> joiner, final boolean outer,
+            final Optional<String> outerJoinWindowName, final TimeTrackerSupplier sharedTimeTrackerSupplier) {
```
Change this back? I think for Kafka streams, the convention is to align with first param's indentation

VictorvandenHoven replied:
changes reverted

lihaosky commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
```diff
@@ -94,7 +92,8 @@ public Processor<K, V1, K, VOut> get() {
         private Sensor droppedRecordsSensor;
-        private Optional<KeyValueStore<TimestampedKeyAndJoinSide<K>, LeftOrRightValue<V1, V2>>> outerJoinStore = Optional.empty();
+        private Optional<KeyValueStore<TimestampedKeyAndJoinSide<K>, LeftOrRightValue<V1, V2>>> outerJoinStore = Optional
+                .empty();
```
revert this change? Does it cause checkstyle problem previously?

VictorvandenHoven replied:
changes reverted

lihaosky commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
```diff
@@ -104,20 +103,16 @@ public void init(final ProcessorContext<K, VOut> context) {
             final StreamsMetricsImpl metrics = (StreamsMetricsImpl) context.metrics();
-            droppedRecordsSensor = droppedRecordsSensor(Thread.currentThread().getName(), context.taskId().toString(), metrics);
+            droppedRecordsSensor = droppedRecordsSensor(Thread.currentThread().getName(), context.taskId().toString(),
+                    metrics);
```
ditto

VictorvandenHoven replied:
changes reverted

lihaosky commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
```diff
@@ -104,20 +103,16 @@ public void init(final ProcessorContext<K, VOut> context) {
-                    )
-                );
+                sharedTimeTracker.setEmitInterval(StreamsConfig.InternalConfig.getLong(context.appConfigs(),
+                        EMIT_INTERVAL_MS_KSTREAMS_OUTER_JOIN_SPURIOUS_RESULTS_FIX, 1000L));
```
ditto

VictorvandenHoven replied:
changes reverted

lihaosky commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
```diff
@@ -134,29 +129,24 @@ public void process(final Record<K, V1> record) {
+                            null));
+
+                    context().forward(record.withValue(joiner.apply(record.key(), record.value(), otherRecord.value))
+                            .withTimestamp(Math.max(inputRecordTimestamp, otherRecordTimestamp)));
```
ditto

VictorvandenHoven replied:
changes reverted

lihaosky commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
```diff
@@ -165,40 +155,52 @@ public void process(final Record<K, V1> record) {
-                            TimestampedKeyAndJoinSide.make(isLeftSide, record.key(), inputRecordTimestamp),
-                            LeftOrRightValue.make(isLeftSide, record.value())));
+                                TimestampedKeyAndJoinSide.make(isLeftSide, record.key(), inputRecordTimestamp),
+                                LeftOrRightValue.make(isLeftSide, record.value())));
```
ditto

VictorvandenHoven replied:
changes reverted

lihaosky reviewed (commented):
LGTM overall! Thanks! Can you rebase to trunk to resolve conflict? Also there are a lot of unnecessary format changes, can you revert those?

guozhangwang commented:
Also made a very quick pass, and I think the fix is spot on. It would be great to get this merged sooner.

VictorvandenHoven committed:
KAFKA-15417  more compliant to kafka-streams layout conventions

VictorvandenHoven committed:
Merge remote-tracking branch 'origin/trunk' into
KAFKA-15417-flip-joinSpuriousLookBackTimeMs

VictorvandenHoven committed:
KAFKA-15417  re-formatted the layout to the previous version

lihaosky commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamOuterJoinTest.java:
```diff
@@ -791,7 +791,7 @@ public void shouldNotEmitLeftJoinResultForAsymmetricBeforeWindow() {
 
             processor.checkAndClearProcessResult(
-                new KeyValueTimestamp<>(1, "null+a1", 102L)
+                    new KeyValueTimestamp<>(1, "null+a1", 102L)
```
revert this?

VictorvandenHoven replied:
Done

lihaosky commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamOuterJoinTest.java:
```diff
@@ -878,16 +880,18 @@ public void shouldNotEmitLeftJoinResultForAsymmetricAfterWindow() {
 
             processor.checkAndClearProcessResult(
-                new KeyValueTimestamp<>(2, "A2+a2", 201L)
+                    new KeyValueTimestamp<>(2, "A2+a2", 201L)
```
ditto

VictorvandenHoven replied:
Done

lihaosky reviewed (approved):
LGTM except two format problem. cc @mjsax

VictorvandenHoven committed:
KAFKA-15417  formatting to kafka-streams conventions

VictorvandenHoven committed:
KAFKA-15417  Checkstyle: removed a white space

VictorvandenHoven commented:
In javaDoc  of JoinWindows:

`There are three different window configuration supported:
before = after = time-difference
before = 0 and after = time-difference
before = time-difference and after = 0
A join is symmetric in the sense, that a join specification on the first stream returns the same result record as a join specification on the second stream with flipped before and after values.
Both values (before and after) must not result in an "inverse" window, i.e., upper-interval bound cannot be smaller than lower-interval bound.`

I think with this change that **non-symmetric** values for the window configuration can be supported as well now.

lihaosky commented:
@mjsax , @guozhangwang , can we merge this?

VictorvandenHoven commented:
> @mjsax , @guozhangwang , can we merge this?

Since it has been a couple of months, I suppose it will not be merged then?

Can we discuss this?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoinTest.java:
```diff
@@ -436,6 +436,239 @@ public void testRightNonJoinedRecordsAreNeverEmittedByTheRightProcessor() {
+            );
+            
+            // push four items smaller timestamps (out of window) to the secondary stream; 
+            // this should produce four joined items
```
I don't think we should get this output? We closed the first three join windows already and emitted `A0+null`, `A1+null`, and `A2+null`, and thus the first three records from the second stream should be dropped on the floor as "late" records? Only `3:a3 (ts: 999)` can produce output because the fourth join-window is still open.

VictorvandenHoven replied:
Agreed!
Assuming that the (no-)grace period only aplies to the joinwindows on the left part (before=100, after=0, grace=0). 
Then the a3 that arrives on the right part (before=0, after=100, grace=?) is simply too late and the A3 should be emitted with null.
Thus, I think that the output of the second part should be:
`            processor.checkAndClearProcessResult(
                    new KeyValueTimestamp<>(3, "A3+null", 1003L)
            );
`

I think that you should forward only records from the right part if the inputRecordTimeStamp + grace period is not later than streamTime.
What do you think?

mjsax replied:
We filed https://issues.apache.org/jira/browse/KAFKA-16123 in the mean time.

It might be best to merge this PR as-is, and tackle the grace-period issue (which is more generic an independent of this fix) in a follow up PR.

VictorvandenHoven replied:
Ok thanks!

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoinTest.java:
```diff
@@ -436,6 +436,239 @@ public void testRightNonJoinedRecordsAreNeverEmittedByTheRightProcessor() {
+            );
+
+            // push four items with increased timestamps to the secondary stream; 
+            // this should produce four joined item
```
Same as above. I think we should only get a result for `b3`?

VictorvandenHoven commented:
> @mjsax , @guozhangwang , can we merge this?

How long does it normally take to get a reaction?

VictorvandenHoven commented:
Accidently closed the PR, reopening again

VictorvandenHoven committed:
Merge branch 'apache:trunk' into KAFKA-15417-flip-joinSpuriousLookBackTimeMs

VictorvandenHoven committed:
KAFKA-15417 merged code from KAFKA-16123

VictorvandenHoven committed:
KAFKA-15417 merged code from KAFKA-16123: Fix
KStreamKStreamJoinProcessor forwards null records for left/outer joins
unconditionally of the join window

VictorvandenHoven committed:
Merge branch 'KAFKA-15417-flip-joinSpuriousLookBackTimeMs' of https://github.com/VictorvandenHoven/VixKafka.git into KAFKA-15417-flip-joinSpuriousLookBackTimeMs

VictorvandenHoven commented:
Merged the code of [KAFKA-16123](https://issues.apache.org/jira/browse/KAFKA-16123) into this PR.
Everything else left as it was.
All the tests still passed.

mjsax commented:
> Merged the code of [KAFKA-16123](https://issues.apache.org/jira/browse/KAFKA-16123) into this PR.

Why? We are mixing up two ticket if we do this (cf https://github.com/apache/kafka/pull/14426#discussion_r1483677544)

Can you remove those changes? Fixing the grace period should be kept separate to get different commits for different fixes.

florin-akermann commented:
> > All the tests still passed.

What currently has been merged from https://issues.apache.org/jira/browse/KAFKA-16123 into this PR wouldn't solve the general case (non null-key records).

I agree with Matthias, I would keep it separate.

florin-akermann commented:
I now pushed a 'generalized' fix for [KAFKA-16123](https://issues.apache.org/jira/browse/KAFKA-16123)

VictorvandenHoven committed:
Merge branch 'apache:trunk' into KAFKA-15417-flip-joinSpuriousLookBackTimeMs

VictorvandenHoven committed:
Revert "KAFKA-15417 merged code from KAFKA-16123"

This reverts commit 2cf235a186d3b2545a12794c21d7ca91226f8ab4.

VictorvandenHoven committed:
Revert "Revert "KAFKA-15417 merged code from KAFKA-16123""

This reverts commit f00d0f44477cf4277db3ee88f4b9cb3d27e4e408.

VictorvandenHoven committed:
Revert "Revert "Revert "KAFKA-15417 merged code from KAFKA-16123"""

This reverts commit 7e75db71e430de61ae2fa2e6c65e2f81e577ff72.

VictorvandenHoven commented:
So I reverted the code of [KAFKA-16123](https://issues.apache.org/jira/browse/KAFKA-16123).

VictorvandenHoven committed:
Merge branch 'apache:trunk' into KAFKA-15417-flip-joinSpuriousLookBackTimeMs

VictorvandenHoven committed:
KAFKA-15417  changing back the ordering to what is was before

mjsax commented:
@VictorvandenHoven -- it seems `KStreamKStreamIntegrationTest.shouldOuterJoin` fails consistently. Can you take a look?

VictorvandenHoven commented:
> @VictorvandenHoven -- it seems `KStreamKStreamIntegrationTest.shouldOuterJoin` fails consistently. Can you take a look?

Ouch, didn't test that one.

Apparently, the internalProcessorContext.currentSystemTimeMs() behaves differently as expected.
The **MOCK_TIME** advances 10 seconds and the **internalProcessorContext.currentSystemTimeMs()** only advances about 100 ms.
Since the "next time to emit interval" is 1000ms by default, the "next time to emit" will never be reached.

When I change the value of the  "next time to emit interval" to 0,  by adding the following line in de @BeforeEach.
`            streamsConfig.put(InternalConfig.EMIT_INTERVAL_MS_KSTREAMS_OUTER_JOIN_SPURIOUS_RESULTS_FIX, 0L);
`
Then the test succeeds.

I do not understand however, why this test worked before?

Shall I change the test with the EMIT_INTERVAL_MS_KSTREAMS_OUTER_JOIN_SPURIOUS_RESULTS_FIX?

VictorvandenHoven committed:
KAFKA-15417  set
EMIT_INTERVAL_MS_KSTREAMS_OUTER_JOIN_SPURIOUS_RESULTS_FIX to 0

VictorvandenHoven committed:
Merge remote-tracking branch 'origin/trunk' into
KAFKA-15417-flip-joinSpuriousLookBackTimeMs

VictorvandenHoven committed:
KAFKA-15417 remove debug lines

mjsax commented:
Looking into the test, we create `new KafkaStreams(builder.build(streamsConfig), streamsConfig)`, but we don't pass in the mock time object. So KS creates it's own `Time` object, so it's decoupled...

But yes, setting `streamsConfig.put(InternalConfig.EMIT_INTERVAL_MS_KSTREAMS_OUTER_JOIN_SPURIOUS_RESULTS_FIX, 0L);` is a valid fix, and we do the same thing for others test -- in the end, this config is a perf optimization, but for testing we can disable it by setting it to zero.

> I do not understand however, why this test worked before?

Not 100% sure TBO.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoinTest.java:
```diff
@@ -436,6 +436,239 @@ public void testRightNonJoinedRecordsAreNeverEmittedByTheRightProcessor() {
+            // push four items with increasing timestamps to the primary stream; this should emit null-joined items
+            // w1 = {}
+            // w2 = {}
+            // --> w1 = { 0:B0 (ts: 1000), 1:B1 (ts: 1001), 2:B2 (ts: 1002), 3:B3 (ts: 1003) }
```
Why are we using `B` not `A` for left input?

VictorvandenHoven replied:
Changed B into A

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoinTest.java:
```diff
@@ -436,6 +436,239 @@ public void testRightNonJoinedRecordsAreNeverEmittedByTheRightProcessor() {
+            final MockApiProcessor<Integer, String, Void, Void> processor = supplier.theCapturedProcessor();
+
+            processor.init(null);
+            // push four items with increasing timestamps to the primary stream; this should emit null-joined items
```
`his should emit [three] null-joined items; the last one does not emit yet`

VictorvandenHoven replied:
modified comment

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoinTest.java:
```diff
@@ -436,6 +436,239 @@ public void testRightNonJoinedRecordsAreNeverEmittedByTheRightProcessor() {
     }
 
+    @Test
+    public void testLeftNonJoinedRecordsWithZeroAfterAreEmitted() {
```
Why do we need this test case? Seems it's fully contained in `testLeftJoinedRecordsWithZeroAfterAreEmitted` below?

VictorvandenHoven replied:
Correct, removed this test.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamOuterJoinTest.java:
```diff
@@ -438,13 +438,13 @@ public void testOrdering() {
-            // by the time they were produced before
+            // push one item to the other window that has a join;
+            // this should produce the joined record first;
+            // then the not-joined record
```
This change to the comment needs to be rolled back, right? We indeed produce the left-null join result first.

VictorvandenHoven replied:
Modified the comments according to the results produced.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamOuterJoinTest.java:
```diff
@@ -884,11 +886,13 @@ public void shouldNotEmitLeftJoinResultForAsymmetricAfterWindow() {
-            // push one item to the first stream; this should produce one full-join item
+            // push one item to the first stream;
+            // this should produce one inner-join item;
+            // and a right-joined item for a3
```
We don't produce output for `a3`

VictorvandenHoven replied:
Removed the line in the comment that says we produce output for a3.

mjsax reviewed (commented):
Went over the test code again in detail. A few more nits... Can merge after addressed.

VictorvandenHoven committed:
KAFKA-15417-flip-joinSpuriousLookBackTimeMs

process review comments

VictorvandenHoven committed:
KAFKA-15417-flip-joinSpuriousLookBackTimeMs 

process review comments

VictorvandenHoven committed:
long time is not a final

mjsax commented:
Thanks for the fix! Merged to `trunk`.

Really appreciate that you did push this through. Was more complicated than expected and took way to long to get finished.

florin-akermann commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
```diff
@@ -230,8 +234,19 @@ private void emitNonJoinedOuterRecords(
+                        } else {
+                            outerJoinRightBreak = true; // there are no more candidates to emit on right-outerJoin-side
+                        }
+                        if (outerJoinLeftBreak && outerJoinRightBreak) {
```
Hi,

Seems like `outerJoinLeftBreak && outerJoinRightBreak` is always false.
Doesn't this break the behavior described in the comment on top of this block? `// Skip next records if window has not closed`?

VictorvandenHoven replied:
Probably these two lines need to be outside the while loop:
```
                    boolean outerJoinLeftBreak = false;
                    boolean outerJoinRightBreak = false;
```

VictorvandenHoven replied:
See: https://github.com/apache/kafka/pull/15510
